src/txorphanage.cpp

```diff
@@ -226,18 +317,64 @@ void TxOrphanage::EraseForBlock(const CBlock& block)
             if (itByPrev == m_outpoint_to_orphan_it.end()) continue;
             for (auto mi = itByPrev->second.begin(); mi != itByPrev->second.end(); ++mi) {
                 const CTransaction& orphanTx = *(*mi)->second.tx;
-                const uint256& orphanHash = orphanTx.GetHash();
-                vOrphanErase.push_back(orphanHash);
+                const uint256& orphan_wtxid = orphanTx.GetWitnessHash();
+                vOrphanErase.push_back(orphan_wtxid);
             }
         }
     }
 
     // Erase orphan transactions included or precluded by this block
     if (vOrphanErase.size()) {
         int nErased = 0;
-        for (const uint256& orphanHash : vOrphanErase) {
-            nErased += EraseTxNoLock(orphanHash);
+        for (const uint256& orphan_wtxid : vOrphanErase) {
+            nErased += EraseTxNoLock(orphan_wtxid);
```

Comment:
```
echo Kv//////////////lJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSUlJSU//////////////////////////////+A////////kEEKJf//ev///////8P///////8AAIH//wkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkpKSkB/wD8///SQP///////////////////////////////////4D///////+QQQol//96////////w/////////8KJf//ev//////////kEEKJf//1v///wP/////kEH///+QQQo1//8BPwAAv67Kf/////8u//////////////////////////////////////////////////////////////////////////96AP/1///////S0P//IP96AP////////+A////////kEEKJf//ev//////////kEEKJf//1v///wP/////kEH///+QQQo1//96/////P////+QQQol/yX//3r//////////wol///W//////////97e3t7e3t7e3t7e3t7e3t7e3t7e/+QQQol//96//////////////8s/yb/ | base64 -d > crash.txt
```

These call `EraseTxNoLock` without doing m_peer_work_set management:
TxOrphanage::EraseForBlock
TxOrphanage::LimitOrphans

This results in fuzzer crashes where it's (still) in the peer's work set, but not counted as an announcer since it's been evicted(then later sent).

Any time `EraseTx{NoLock}` is called we need to make sure that `m_peer_work_set` is appropriately being cleaned.

Two other spots I haven't thought through yet but I suspect need the same fix:
TxDownloadImpl::MempoolAcceptedTx
TxDownloadImpl::MempoolRejectedTx

Reply:
Is it a problem to lazily handle transactions in peer work sets that have been deleted from the orphanage? It's cleaner to `remove_work_from_all_sets` of course, but it iterates through every entry in `m_peer_work_set` so there is a performance difference but the external outcome is the same.

I think erasing proactively in `EraseOrphanOfPeer` is necessary because the tx isn't deleted. Trying to avoid returning the tx in `GetTxToReconsider(peer)` after we've already decided to give up on (tx, peer).

Reply:
> TxOrphanage::EraseForBlock
> TxOrphanage::LimitOrphans

> This results in fuzzer crashes where it's (still) in the peer's work set, but not counted as an announcer since it's been evicted(then later sent).

Question: is it crashing because the tx is being returned from `GetTxToReconsider`? Seems weird as it should return a nullptr if it gets a tx that isn't there anymore...

OOhhhhhhhhh could it be that it's re-added? My guess:
1. `AddTx(tx_child, peer)`
2. `AddChildrenToWorkSet(tx_parent)` -> peer workset contains tx_child
3. `EraseTx(tx_child)` -> peer workset still contains tx_child but we expect orphanage to lazily delete that when we call `GetTxToReconsider`
4. `AddTx(tx_child, other_peer)` (makes no sense but we're fuzzing)
5. `GetTxToReconsider(peer)` returns `tx_child` even though it's not in `orphan_resolution_tracker`

Reply:
> Is it a problem to lazily handle transactions in peer work sets that have been deleted from the orphanage?

Since `GetTxToReconsider` appears to be used in a single spot, and it really doesn't matter that it once was announced, then erased, then reconsidered for a peer, I suppose the fuzz test should just be modified to not crash :)

Reply:
Yeah. Removed the `Assert(HaveTxAndPeer())` bit

Reply:
Removed `Assert(HaveTxAndPeer())` in the fuzzer, and added comments for when we clean up the work sets and when we don't